File: backend/internal/server/routes/payment.go
package routes

import (
	"time"

	"github.com/Wei-Shaw/sub2api/internal/handler"
	"github.com/Wei-Shaw/sub2api/internal/handler/admin"
	basemiddleware "github.com/Wei-Shaw/sub2api/internal/middleware"
	"github.com/Wei-Shaw/sub2api/internal/server/middleware"
	"github.com/Wei-Shaw/sub2api/internal/service"

	"github.com/gin-gonic/gin"
	"github.com/redis/go-redis/v9"
)

// RegisterPaymentRoutes registers all payment-related routes:
// user-facing endpoints, webhook endpoints, and admin endpoints.
func RegisterPaymentRoutes(
	v1 *gin.RouterGroup,
	paymentHandler *handler.PaymentHandler,
	webhookHandler *handler.PaymentWebhookHandler,
	adminPaymentHandler *admin.PaymentHandler,
	jwtAuth middleware.JWTAuthMiddleware,
	adminAuth middleware.AdminAuthMiddleware,
	auditLog middleware.AuditLogMiddleware,
	settingService *service.SettingService,
	panelRateLimiter *middleware.PanelRateLimiter,
	redisClient *redis.Client,
) {
	// --- User-facing payment endpoints (authenticated) ---
	authenticated := v1.Group("/payment")
	authenticated.Use(gin.HandlerFunc(jwtAuth))
	authenticated.Use(middleware.BackendModeUserGuard(settingService))
	// 面板全局按用户限流
	authenticated.Use(panelRateLimiter.Global())
	{
		authenticated.GET("/config", paymentHandler.GetPaymentConfig)
		authenticated.GET("/checkout-info", paymentHandler.GetCheckoutInfo)
		authenticated.GET("/plans", paymentHandler.GetPlans)
		authenticated.GET("/limits", paymentHandler.GetLimits)
		authenticated.GET("/refunds/overview", paymentHandler.GetAccountRefundOverview)
		authenticated.POST("/refunds/lock", paymentHandler.LockAccountRefund)
		authenticated.POST("/refunds/donate", paymentHandler.DonateAccountRefund)

		orders := authenticated.Group("/orders")
		{
			orders.POST("", paymentHandler.CreateOrder)
			orders.POST("/verify", paymentHandler.VerifyOrder)
			orders.GET("/my", paymentHandler.GetMyOrders)
			orders.GET("/:id", paymentHandler.GetOrder)
			orders.POST("/:id/cancel", paymentHandler.CancelOrder)
			orders.GET("/refund-eligible-providers", paymentHandler.GetRefundEligibleProviders)
		}
	}

	// --- Public payment endpoints (no auth) ---
	// Signed resume-token recovery is the preferred public lookup path.
	// The legacy anonymous out_trade_no verify endpoint remains available as a
	// persisted-state compatibility path for staggered upgrades.
	public := v1.Group("/payment/public")
	{
		public.POST("/orders/verify", paymentHandler.VerifyOrderPublic)
		public.POST("/orders/resolve", paymentHandler.ResolveOrderPublicByResumeToken)
		public.GET("/refund-donations", paymentHandler.ListAccountRefundDonations)
	}

	// 账户被锁定后普通 JWT 会失效，退款专用会话只能访问当前清退流程。
	refundSession := v1.Group("/payment/refunds")
	refundSession.Use(panelRateLimiter.Global())
	{
		// 恢复入口与登录同级失败关闭限流，防止成为密码爆破旁路。
		rateLimiter := basemiddleware.NewRateLimiter(redisClient)
		refundSession.POST("/session/restore", rateLimiter.LimitWithOptions("account-refund-session-restore", 10, time.Minute, basemiddleware.RateLimitOptions{
			FailureMode: basemiddleware.RateLimitFailClose,
		}), gin.HandlerFunc(auditLog), paymentHandler.RestoreAccountRefundSession)
		refundSession.GET("/:refund_id", paymentHandler.GetAccountRefund)
		refundSession.POST("/:refund_id/confirm", paymentHandler.ConfirmAccountRefund)
		refundSession.POST("/:refund_id/donate", paymentHandler.DonateLockedAccountRefund)
		refundSession.POST("/:refund_id/cancel", paymentHandler.CancelAccountRefund)
	}

	// --- Webhook endpoints (no auth) ---
	webhook := v1.Group("/payment/webhook")
	{
		// EasyPay sends GET callbacks with query params
		webhook.GET("/easypay", webhookHandler.EasyPayNotify)
		webhook.POST("/easypay", webhookHandler.EasyPayNotify)
		webhook.POST("/alipay", webhookHandler.AlipayNotify)
		webhook.POST("/wxpay", webhookHandler.WxpayNotify)
		webhook.POST("/stripe", webhookHandler.StripeWebhook)
		webhook.POST("/airwallex", webhookHandler.AirwallexWebhook)
	}

	// --- Admin payment endpoints (admin auth) ---
	adminGroup := v1.Group("/admin/payment")
	adminGroup.Use(gin.HandlerFunc(adminAuth))
	adminGroup.Use(gin.HandlerFunc(auditLog))
	adminGroup.Use(middleware.AdminComplianceGuard(settingService))
	{
		// Dashboard
		adminGroup.GET("/dashboard", adminPaymentHandler.GetDashboard)

		// Config
		adminGroup.GET("/config", adminPaymentHandler.GetConfig)
		adminGroup.PUT("/config", adminPaymentHandler.UpdateConfig)

		// Orders
		adminOrders := adminGroup.Group("/orders")
		{
			adminOrders.GET("", adminPaymentHandler.ListOrders)
			adminOrders.GET("/:id", adminPaymentHandler.GetOrderDetail)
			adminOrders.POST("/:id/cancel", adminPaymentHandler.CancelOrder)
			adminOrders.POST("/:id/retry", adminPaymentHandler.RetryFulfillment)
			adminOrders.POST("/:id/refund", adminPaymentHandler.ProcessRefund)
			adminOrders.POST("/:id/refund/query", adminPaymentHandler.QueryAndFinalizeRefund)
		}

		// Subscription Plans
		plans := adminGroup.Group("/plans")
		{
			plans.GET("", adminPaymentHandler.ListPlans)
			plans.POST("", adminPaymentHandler.CreatePlan)
			plans.PUT("/:id", adminPaymentHandler.UpdatePlan)
			plans.DELETE("/:id", adminPaymentHandler.DeletePlan)
		}

		// 充值赠送活动
		rechargeBonusCampaigns := adminGroup.Group("/recharge-bonus-campaigns")
		{
			rechargeBonusCampaigns.GET("", adminPaymentHandler.ListRechargeBonusCampaigns)
			rechargeBonusCampaigns.POST("", adminPaymentHandler.CreateRechargeBonusCampaign)
			rechargeBonusCampaigns.PUT("/:id", adminPaymentHandler.UpdateRechargeBonusCampaign)
			rechargeBonusCampaigns.DELETE("/:id", adminPaymentHandler.DeleteRechargeBonusCampaign)
		}

		// Provider Instances
		providers := adminGroup.Group("/providers")
		{
			providers.GET("", adminPaymentHandler.ListProviders)
			providers.POST("", adminPaymentHandler.CreateProvider)
			providers.PUT("/:id", adminPaymentHandler.UpdateProvider)
			providers.DELETE("/:id", adminPaymentHandler.DeleteProvider)
		}
	}
}

File: frontend/src/api/payment.ts
/**
 * User Payment API endpoints
 * Handles payment operations for regular users
 */

import { apiClient } from './client'
import type {
  PaymentConfig,
  SubscriptionPlan,
  MethodLimitsResponse,
  CheckoutInfoResponse,
  CreateOrderRequest,
  CreateOrderResult,
  PaymentOrder,
  AccountRefundRecord,
} from '@/types/payment'
import type { BasePaginationResponse } from '@/types'

export interface PublicOrderVerifyResult {
  out_trade_no: string
  status: string
  paid: boolean
  created_at: string
  expires_at: string
}

export const paymentAPI = {
  /** Get payment configuration (enabled types, limits, etc.) */
  getConfig() {
    return apiClient.get<PaymentConfig>('/payment/config')
  },

  /** Get available subscription plans */
  getPlans() {
    return apiClient.get<SubscriptionPlan[]>('/payment/plans')
  },

  /** Get all checkout page data in a single call */
  getCheckoutInfo() {
    return apiClient.get<CheckoutInfoResponse>('/payment/checkout-info')
  },

  /** Get payment method limits and fee rates */
  getLimits() {
    return apiClient.get<MethodLimitsResponse>('/payment/limits')
  },

  /** Create a new payment order */
  createOrder(data: CreateOrderRequest) {
    return apiClient.post<CreateOrderResult>('/payment/orders', data)
  },

  /** Get current user's orders */
  getMyOrders(params?: { page?: number; page_size?: number; status?: string }) {
    return apiClient.get<BasePaginationResponse<PaymentOrder>>('/payment/orders/my', { params })
  },

  /** Get a specific order by ID */
  getOrder(id: number) {
    return apiClient.get<PaymentOrder>(`/payment/orders/${id}`)
  },

  /** Cancel a pending order */
  cancelOrder(id: number) {
    return apiClient.post(`/payment/orders/${id}/cancel`)
  },

  /** Verify order payment status with upstream provider */
  verifyOrder(outTradeNo: string) {
    return apiClient.post<PaymentOrder>('/payment/orders/verify', { out_trade_no: outTradeNo })
  },

  /** Legacy-compatible public order lookup by out_trade_no */
  verifyOrderPublic(outTradeNo: string) {
    return apiClient.post<PublicOrderVerifyResult>('/payment/public/orders/verify', { out_trade_no: outTradeNo })
  },

  /** Resolve an order from a signed resume token without auth */
  resolveOrderPublicByResumeToken(resumeToken: string) {
    return apiClient.post<PublicOrderVerifyResult>('/payment/public/orders/resolve', { resume_token: resumeToken })
  },

  /** Get provider instance IDs that allow user refund */
  getRefundEligibleProviders() {
    return apiClient.get<{ provider_instance_ids: string[] }>('/payment/orders/refund-eligible-providers')
  },

  /** Get the authoritative account-wide refund quote. */
  getAccountRefundOverview() {
    return apiClient.get<AccountRefundRecord>('/payment/refunds/overview')
  },

  /** Lock the account and wait for all billable work to settle. */
  lockAccountRefund(quoteHash: string) {
    return apiClient.post<AccountRefundRecord>('/payment/refunds/lock', { quote_hash: quoteHash })
  },

  /** Restore only the dedicated session for an already locked account refund. */
  restoreAccountRefundSession(email: string, password: string, totpCode: string) {
    return apiClient.post<AccountRefundRecord>('/payment/refunds/session/restore', {
      email,
      password,
      totp_code: totpCode,
    }, { headers: { 'X-Skip-Auth': '1' } })
  },

  /** Read a locked refund with its dedicated session credential. */
  getAccountRefund(refundId: string, sessionToken: string) {
    return apiClient.get<AccountRefundRecord>(`/payment/refunds/${refundId}`, {
      headers: refundSessionHeaders(sessionToken),
    })
  },

  /** Confirm the immutable final quote. */
  confirmAccountRefund(refundId: string, quoteHash: string, sessionToken: string) {
    return apiClient.post<AccountRefundRecord>(`/payment/refunds/${refundId}/confirm`, { quote_hash: quoteHash }, {
      headers: refundSessionHeaders(sessionToken),
    })
  },

  /** Cancel before any gateway submission and restore the previous account state. */
  cancelAccountRefund(refundId: string, sessionToken: string) {
    return apiClient.post<AccountRefundRecord>(`/payment/refunds/${refundId}/cancel`, {}, {
      headers: refundSessionHeaders(sessionToken),
    })
  }
}

function refundSessionHeaders(sessionToken: string) {
  return {
    'X-Refund-Session': sessionToken,
    'X-Skip-Auth': '1',
  }
}
